### PR TITLE
CI: major rework for image generations of GHA

### PR DESCRIPTION
**Pull request by vicentebolea**
CI: major rework for image generations of Github Actions builds.

- It adds Kokkos CUDA.
- It sets the CMAKE version to 3.22.4.
- It leverages in Spack Environments to install the dependencies needed for the builds
- In the CUDA image, we install CUDA in Spack rather than using the CUDA rockylinux image.
- It reduces the amount of intermediate images.
- It introduces pyexpand templates to easily modify packages and versions per each build in the future wihouth changing the dockerfiles and shell scrips.

### CODE REVIEW

File: scripts/ci/images-v2/build-cuda.sh [hunk 1]
```diff
@@ -32,7 +32,10 @@ if ! build_squash \
     ornladios/adios2:ci-spack-el8-cuda-gcc8-base \
     ornladios/adios2:ci-spack-el8-cuda-serial \
     Dockerfile.ci-spack-el8-leaf \
-    "--build-arg=COMPILER_IMG_BASE=${build_conf} --build-arg=COMPILER_SPACK_ID=gcc --build-arg=CUDA_VARIANT=+cuda"
+    "--build-arg=COMPILER_IMG_BASE=${build_conf}
+     --build-arg=COMPILER_SPACK_ID=gcc
+     --build-arg=CUDA_VARIANT=+cuda
+     --build-arg=EXTRA_SPACK_PACKAGES=kokkos@3.6.01%gcc+cuda+wrapper"
```
Comment (anagainaru) [inline scripts/ci/images-v2/build-cuda.sh:38]: I did not have time to reproduce this to test my theory, but if you are using the wrapper for kokkos, it needs to be used for adios as well. How is adios build in the CI?

**Review by anagainaru — COMMENTED** [review]
You should be able to load kokkos without the wrapper. Did you say you were getting errors?

I haven't tested yet, but ADIOS2 and Kokkos need to both be build with the nvcc wrapper or both without.

File: scripts/ci/images-v2/build-cuda.sh (hunk 1 above)
Comment (vicentebolea) [reply scripts/ci/images-v2/build-cuda.sh:38]: Here you can change the flags for the CUDA build

https://github.com/ornladios/ADIOS2/blob/master/scripts/ci/cmake-v2/ci-el8-cuda-serial.cmake

Comment (vicentebolea) [reply scripts/ci/images-v2/build-cuda.sh:38]: > You should be able to load kokkos without the wrapper. Did you say you were getting errors?

If I do a `kokkos@3.6.01%gcc+cuda+wrapper` with spack 0.18.1

File: source/adios2/CMakeLists.txt [hunk 2]
```diff
@@ -120,7 +120,7 @@ if(ADIOS2_HAVE_CUDA)
     OUTPUT_NAME adios2${ADIOS2_LIBRARY_SUFFIX}_core_cuda
     )
   target_compile_features(adios2_core_cuda PRIVATE cuda_std_14)
-  target_link_libraries(adios2_core PRIVATE adios2_core_cuda CUDA::cudart CUDA::cuda_driver)
```
Comment (vicentebolea) [inline source/adios2/CMakeLists.txt:123]: @anagainaru I have removed this dependency as we did not use it so far and it is difficult to provide through the CI.

**Comment by vicentebolea** [comment]
Merging without review since these are CI changes about the images and many iterations are expected. The images are already deployed and working

**Pull request merged**